### ISSUE TITLE
Unable to send a command with more than 64 characters

### ISSUE THREAD

**Issue by AGlass0fMilk** [ARMmbed/mbed-os-bluetooth-integration-testsuite#27]
As mentioned in another issue, I am attempting to write integration tests using this framework with greentea for a UARTService implementation.

The first tests I am writing are simply loopback tests where the blecli-app is used to write data to the greentea DUT running a `GattServer` with a `UARTService`.

Currently, the writes are limited to the current MTU of the connection, which defaults to 20 bytes.

I found that the tests were timing out when the size of the random payload went above 13 bytes (oooh spooky :jack_o_lantern:). The timeout was evidently coming from the ble-cliapp serial connection as the logs showed this:

```
[1603472211.99][mbed_ble_test_suite.common.device]0x7f80291032b0: Sending command to client: "gattClient writeWithoutResponse 1 13 645F7A3275BDD46E4AE5E3ADE6AA"
[1603472211.99][mbed_ble_test_suite.common.serial_device]-->|0x7f80291032b0| gattClient writeWithoutResponse 1 13 645F7A3275BDD46E4AE5E3ADE6AA
[1603472212.99][mbed_ble_test_suite.common.device]0x7f80291032b0: Waiting for "retcode: 0" string... Timeout in 29 s
```

When the payload is 13 bytes and under, the logs look like this:

```
[1603472268.45][mbed_ble_test_suite.common.device]0x7fd869e2c2b0: Sending command to client: "gattClient writeWithoutResponse 1 13 B1BDDE7DA78EC881B0C27A89D6"
[1603472268.45][mbed_ble_test_suite.common.serial_device]-->|0x7fd869e2c2b0| gattClient writeWithoutResponse 1 13 B1BDDE7DA78EC881B0C27A89D6
[1603472268.46][mbed_ble_test_suite.common.serial_device]<--|0x7fd869e2c2b0| {"status": 0}
[1603472268.46][mbed_ble_test_suite.common.serial_device]<--|0x7fd869e2c2b0| retcode: 0
[1603472268.52][mbed_ble_test_suite.common.serial_device]<--|0x7fd869e2c2b0| 
[1603472268.53][mbed_ble_test_suite.common.serial_device]<--|0x7fd869e2c2b0| <<< {"connHandle": 1,"handle": 15,"type": "BLE_HVX_NOTIFICATION","data": "B1BDDE7DA78EC881B0C27A89D6"}
[1603472268.53][TEST][INF] received data: B1BDDE7DA78EC881B0C27A89D6
```

I will point a few things:

1. The hang is from the ble-cliapp. This is confirmed by the fact that the "above-13-bytes" case has no status/result code in the log.
2. If you inspect two commands, you will notice that the 13-byte command has 63 characters while the 14-byte command has 65 characters.

Due to point 2 my assumption is that somewhere in the code there is a buffer limited to 64 bytes that is causing this issue...

I appreciate any hints while I look for the cause.

**Comment by AGlass0fMilk** [comment]
I actually now suspect this to be a problem with the JLink OB that is handling the USB<->UART passthrough.

Freezing the ble-cliapp after attempting the longer-than-13-bytes write shows some corruption at the beginning/end of the buffer:

![Screenshot from 2020-10-23 13-34-43](https://user-images.githubusercontent.com/350500/97035602-93f93e80-1534-11eb-8098-3c588a71c389.png)

**Comment by AGlass0fMilk** [comment]
Confirming now that adding a non-zero `serial_interbyte_delay` resolves this issue... so it seems that the JLink OB serial passthrough cannot handle transfers larger than 64-bytes at a time.

It could be a limitation of the USB endpoint type being used as well.

@pan- Have you encountered this issue before?

I'll see if we can get any response from Segger.

**Comment by AGlass0fMilk** [comment]
See Segger forum post here (pending moderator approval):

https://forum.segger.com/index.php/Thread/7578-JLinkOB-Serial-Passthrough-Payload-Size-Limited-to-64-Bytes/

**Comment by pan-** [comment]
Yes we encountered a similar issue before. Four years ago, on NRF51! At the time Mbed OS serial support was very poor and I had to create the first `CircularBuffer` to alleviate the issue. That primitive was latter picked up and upstreamed to  Mbed OS .

A similar issue was reported few days ago on the latest ST targets running ST Link. It **might** not be related to JLink. I would like to update the code using `mbed::BufferedSerial` to see if it improves the situation. I will issue a PR latter today.

**Comment by AGlass0fMilk** [comment]
I've seen issues with Jlink OB before where if I can actually get the whole board to crash (and cause a USB re-enumeration) just through serial traffic alone. I'm wondering if maybe that's related.

It could be a limitation of the USB endpoint/transfer types used by the JLink and/or CDC class.

It might be a good idea to rework the pyserial implementation to introduce a `max_transfer_size` option. Then you could make sure to break transfers up into multiple smaller transfers of at most that size without introducing a delay between each byte.

**Comment by pan-** [comment]
That's good feedback. 

It is a tricky subject many boards exhibit look alike issues but the actual cause depends on the board/interface firmware. The `serial_interbyte_delay` was added to help ST with testing as on some of their boards bytes are lost. 
I can add a `max_transfer_size` option to the test suite, I will also make the `BufferedSerial` change. I prefer to have the maximum area covered here so if there's an issue with a new config there's at least a workaround available.
